Support DEFAULT column values.

petermattis opened the pull request:
Added support for default value expressions in column definitions. Added
support for default value expressions to ColumnDescriptor. Display
default value expressions in "SHOW COLUMNS" statements. Use DEFAULT
expressions on INSERT and UPDATE when the DEFAULT marker is used or a
value is not specified.

Fixes #2466.
See #2513.

petermattis committed:
Support DEFAULT column values.

Added support for default value expressions in column definitions. Added
support for default value expressions to ColumnDescriptor. Display
default value expressions in "SHOW COLUMNS" statements. Use DEFAULT
expressions on INSERT and UPDATE when the DEFAULT marker is used or a
value is not specified.

Fixes #2466.
See #2513.

tamird commented on cli/sql_util_test.go:
```diff
@@ -93,13 +93,13 @@ OK
++------------+----------+------+---------+
+|   Field    |   Type   | Null | Default |
++------------+----------+------+---------+
+| "parentID" | "INT"    | true | ""      |
```
shouldn't this default be NULL?

petermattis replied:
Yes, though `NULL` prints as `<nil>` with our cli output routines.

tamird commented on sql/insert.go:
```diff
@@ -53,6 +55,17 @@ func (p *planner) Insert(n *parser.Insert) (planNode, error) {
+		if _, ok := colIDtoRowIndex[col.ID]; ok {
+			continue
+		}
+		if col.DefaultExpr != "" {
```
I think `DefaultExpr` is going to have to be nullable - it seems impossible to have an empty string default with this representation.

petermattis replied:
Yeah, that's a good point. Will change.

petermattis committed:
Make ColumnDescriptor.default_expr nullable.

tamird commented on sql/testdata/default:
```diff
@@ -0,0 +1,72 @@
+query IBB
+SELECT a, b >= now(), c >= 0.0 FROM t
+----
+42 true true
```
shouldn't there be two rows here?

petermattis replied:
You probably missed the `TRUNCATE TABLE`. Necessary or we wouldn't be able to insert a row with `a = 42` again.

tamird replied:
I sure did.

tamird commented on sql/testdata/default:
```diff
@@ -0,0 +1,72 @@
+SELECT a, b >= now(), c >= 0.0 FROM t WHERE a = 2
+----
+2 true true
+
```
remove this

petermattis replied:
Done.

tamird commented:
LGTM

vivekmenezes commented on sql/testdata/default:
```diff
@@ -0,0 +1,72 @@
+INSERT INTO t VALUES (DEFAULT, DEFAULT, DEFAULT)
+
+query IBB
+SELECT a, b >= now(), c >= 0.0 FROM t
```
how is b > now() ?

petermattis replied:
Heh, that's a very good question. I'm not sure what is going on, but obviously this is wrong.

petermattis replied:
Heh, `>=` and `<=` comparisons were broken for timestamps.

vivekmenezes replied:
thanks for fixing it!

bdarnell commented on sql/insert.go:
```diff
@@ -186,3 +224,94 @@ func (p *planner) processColumns(tableDesc *TableDescriptor,
+	}
+
+	sql := buf.String()
+	stmts, err := parser.Parse(sql, parser.Syntax(p.session.Syntax))
```
This makes my sql-injection sense twitch. Can we not build up the AST directly instead of building a string and parsing it? 

`p.session.Syntax` is incorrect: the default expressions must be parsed in the language context in which they were defined, not the context of the session that causes them to be evaluated. (Unless we can guarantee that defaults are limited to expressions that are parsed identically in all syntaxes, in which case it would still be better to just hard-code one syntax here instead of using the session)

petermattis replied:
I don't see an easy way to build the AST directly. What would I store in `ColumnDescriptor` in order to build the AST for `now()`? Or, more interestingly, `extract(millisecond from now())`? While the expression is variable free, it can be arbitrarily complex.

You're correct about the use of `p.session.Syntax`. We actually only use the syntax setting for scanning tokens and always output "traditional" syntax. So this should probably be `parser.ParseTraditional`.

tamird replied:
I wouldn't be too concerned about SQL-injection here; we're only parsing, not executing.

bdarnell replied:
I was thinking about building up the select statement, and by building the AST directly I meant to parse each expression individually and then build a select statement from that, without ever appending commas to a buffer. (of course that wouldn't actually make sense here because you don't care about the select statement; you're just using it as a way to get to the expression parser). Could we have a `parser.ParseExpr` function to use here?

petermattis replied:
We're evaluating the parsed expressions, isn't that executing? On the safety side, the expressions have to be free of variables (checked via `parser.ContainsVars`). There isn't much you can do with an expression that doesn't contain variables.

tamird replied:
> We're evaluating the parsed expressions, isn't that executing?

Well, not really. It wouldn't be possible to inject `DEFAULT 1; DROP TABLE students;`, for instance.

bdarnell replied:
We check `parser.ContainsVars` when executing the DDL statements, but we don't repeat this validation when evaluating the defaults. If an expression parses differently between the different syntax modes then the use of `p.session.Syntax` could have allowed a variable reference to sneak through. It would probably good to repeat that check in any case to guard against any potential bugs in our parse-stringify-parse cycle.

We also mentioned subqueries as something that shouldn't be allowed in defaults, but I don't see any enforcement or testing of that in this change.

SQL injection admittedly isn't a major concern here - any attacker who can get something tricky installed as a default expression has probably won the game and doesn't need any further holes.

petermattis replied:
Ok, I'll tighten this up. Subqueries would be fine as long as they don't contain a variable. For example, the default expression could be `(SELECT 1)`. But `(SELECT * FROM t)` contains a variable and would be reject. That said, we don't actually fill in subqueries here, so they wouldn't work. I'll add some code to reject subqueries at column definition time.

petermattis replied:
We can add a `parser.ParseExpr` function, but it would simply be sugar for `SELECT <expr>`.

petermattis replied:
Ok, added `parser.ParseExpr` and re-performed the `ContainsVars` check when the expression is re-parsed. I added a TODO about checking for subqueries. This PR is large enough already.

@tamird Let me know if you feel the urge to replace the current expression parsing in tests with `parser.ParseExpr`. That would be a nice bit of cleanup and I'll get to it eventually if you don't.

petermattis committed:
Fix >= and <= comparisons for timestamps.

Fix broken default tests.

tamird commented on cli/sql_util_test.go:
```diff
@@ -74,15 +74,15 @@ OK
-		{`"parentID"`, `"INT"`, `true`},
-		{`"name"`, `"STRING"`, `true`},
-		{`"id"`, `"INT"`, `true`},
+		{`"parentID"`, `"INT"`, `true`, `<nil>`},
```
looking at this again, we should print `NULL` as `NULL`. Seems like that would go in `sqlRowsToStrings`

petermattis replied:
Agreed. I'll fix in a subsequent change.

tamird replied:
:ok_hand:

petermattis committed:
Always parse default expressions in "traditional" syntax.

The default expressions could have been declared using "modern" syntax,
but they are regenerated using "traditional" syntax before being stored
in the column descriptor.

petermattis committed:
Added parser.ParseExpr and parse the default expressions one at a time.

petermattis committed:
Add TODO about rejecting default expressions containing subqueries.
